### DISCUSSION TITLE
tutorials/physics/ray-casting

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#100] [User-contributed notes]
# tutorials/physics/ray-casting

Introduction: One of the most common tasks in game development is casting a ray (or custom shaped object) and checking what it hits. This enables complex behaviors, AI, etc. to take place. This tut...

https://docs.godotengine.org/en/stable/tutorials/physics/ray-casting.html

**Comment by LordMcMutton** [comment]
Be sure that the script you're using this code in extends the proper Node type: "extends Node2D" for 2D Raycasts, and "extends Node3D" for 3D Raycasts.

**Reply by heero7** [reply]
Saved me in 2024!

**Reply by heero7** [reply]
FYI, my issue was that the parent node was a `Node` instead of a `Node2D`

**Comment by steelywing** [comment]
Currently C# `IntersectRay()` API is a lot slower than `RayCast` node, use `RayCast` node if it is possible.
Detail: see [Godot is not the new Unity](https://sampruden.github.io/posts/godot-is-not-the-new-unity/).

**Reply by steelywing** [reply]
[Here](https://gist.github.com/reduz/cb05fe96079e46785f08a79ec3b0ef21) is Godot contributor Reduz explain

**Reply by SohamTilekar** [reply]
Is the Godot GD script `intersect_ray` is faster or `RayCast` node is faster.

**Comment by n-l** [comment]
A note about the example in the ***3D ray casting from screen*** section.

**If you're developing editor plugins** and need object picking in the editor, note that the `position` property of `InputEventMouseButton` returns the position relative to the window (excluding the title bar), **not** the position relative to the top-left of the 3D viewport.

Instead, as argument to `Camera3D.project_ray_origin()` and `Camera3D.project_ray_normal()`, you need to either use:
```gdscript
EditorInterface.get_editor_viewport_3d(idx).get_mouse_position()
```
or subtract the position of the viewport's parent control from `event.position` (the viewport itself does not have a `position` property).
```gdscript
event.position - EditorInterface.get_editor_viewport_3d(idx).get_parent().global_position
```

**Comment by ghost** [comment]
The 3D examples here show only how to get it working with colliders. However, often it is easier and more efficient to query where the raycast would intercept the zero-plane. You can do this as follows:

```gdscript
var mouse_position := get_window().get_mouse_position()
var camera         := get_window().get_viewport().get_camera_3d()
var plane          := Plane.PLANE_XZ

var intersect : Variant = plane.intersects_ray(
    camera.project_ray_origin(mouse_position),
    camera.project_ray_normal(mouse_position)
)
# Prints either null or the Vector3 where the raycast intersects the zero-plane:
print(intersect)
```

If needed, you can shift the zero-plane up and down by changing the `d` parameter of the `plane`.

**Reply by oddikat** [reply]
is this for a top down view/camera?

**Reply by ghost** [reply]
Hi @oddikat this should work regardless of camera perspective as long as the mouse cursor is over the ground. The most common use would be a top-down view, but you can also use this for a first person build mechanic if you wish.

**Reply by sebastiancarlsson** [reply]
Hello @abstractionfactory, thanks for your comment, it helped me. However I think this is not correct:
> If needed, you can shift the zero-plane up and down by changing the y parameter of the plane.

I tried to do this but I think the `y` parameter represents the `y` normal of the plane, not it's `y` position. I'm not sure if it's possible to shift the position.

I would like to use a plane at a positive `y` position but I couldn't make it work. I was able to do a workaround where I simply shifted everything in my 3d world "down" on the y-axis. Preferrably I would like to not have to use this workaround so if you have a different solution or if I misunderstood something, please let me know. Thank you!

**Reply by sebastiancarlsson** [reply]
I figured it out 30 seconds after I wrote the above. If you want to shift the zero-planes' position you can create a plane like this:
```gdscript
var plane := Plane(0, 1, 0, 100)
```
This example will create a Plane with a positive Y normal and a distance of 100 units from "zero".
https://docs.godotengine.org/en/stable/classes/class_plane.html#class-plane-constructor-plane

**Comment by RaffyTaffy95** [comment]
Hey, wanted to mention that the projecting from camera example might be wrong.
When using PhysicsRayQueryParameters both 'from' and 'to' are global coordinates - that's why, if we have a direction (local), we need to add origin to it, ie. var to = from + direction * length

However, when projecting directly from camera, and using Perspective view in the camera, adding the origin results in more and more offset/raycast being skewed the more object changes its global_position.

Here is my code snippet for moving a "3d cursor" around that is raycasted from camera and intersected with a flat plane that faces upwards and is at the level of the 'player character' (Y-position). Take notice no "origin" being added to "to" var.
Also, coincidentally, project_ray_origin() is identical to camera.global_position in Perspective view.

```gdscript
var mouse_pos = get_viewport().get_mouse_position()
var camera = get_viewport().get_camera_3d()
var from = camera.project_ray_origin(mouse_pos)
var to = camera.project_ray_normal(mouse_pos) * 1000
var drop_plane = Plane(Vector3.UP, global_position.y)
var cursor_3d_pos = drop_plane.intersects_ray(from, to)
```

**Reply by westin813** [reply]
no the example is correct plane.intersect() takes a direction as it's second parameter not the end of a ray

**Comment by FeralBytes** [comment]
What is this metadata: Variant() # metadata of collider? I tried adding metadata to the CollisionShape3D and the StaticBody3D but the result from intersect_ray() never returns metadata as one of the keys.

**Comment by MrPanda05** [comment]
If someone needs help casting the collider into an actual class in C#, use this

```
hit["collider"].AsGodotObject() is CollisionObject3D coli
```

it doesn't need to be an CollisionObject3D directly, but it needs to inherit CollisionObject either 2D or 3D, such an Area or staticBody and so on

More info: https://stackoverflow.com/questions/77262145/godot-4-dictionary-issues-with-physicasrayqueryparameters2d-c-sharp

**Comment by iganinja** [comment]
As a note, the result of the raycast in 3D also includes `face_index` entry:

```
{
   position: Vector2 # point in world space for collision
   normal: Vector2 # normal in world space for collision
   collider: Object # Object collided or null (if unassociated)
   collider_id: ObjectID # Object it collided against

   face_index: int # Collided object's face index, if it's a trimesh?

   rid: RID # RID it collided against
   shape: int # shape index of collider
   metadata: Variant() # metadata of collider
}
```

**Comment by Rauchwerfer** [comment]
`PhysicsDirectSpaceState3D::_intersect_ray` does not return metadata (from `godot/servers/physics_3d
/physics_server_3d.cpp`, at `PhysicsDirectSpaceState3D::_intersect_ray`):

```
Dictionary d;
d["position"] = result.position;
d["normal"] = result.normal;
d["face_index"] = result.face_index;
d["collider_id"] = result.collider_id;
d["collider"] = result.collider;
d["shape"] = result.shape;
d["rid"] = result.rid;

return d;
```
Same for implementations of `intersect_ray` in `JoltPhysicsDirectSpaceState3D` at `godot/modules/jolt_physics/spaces
/jolt_physics_direct_space_state_3d.cpp` and `GodotPhysicsDirectSpaceState3D` at `godot/modules/godot_physics_3d
/godot_space_3d.cpp`, no metadata is set in `intersect_ray` calls. It's worth explicitly stating that metadata is absent from the result (although it can be obtained via `"collider"`). 

Also, `"face_index"` will always be -1 unless setting `"physics/jolt_physics_3d/queries/enable_ray_cast_face_index"` is enabled in the project settings (in version 4.6, Jolt physics is specified as the default physics engine).
